pilot/pkg/config/memory/monitor.go

```diff
@@ -72,32 +79,45 @@ func newBufferedMonitor(store model.ConfigStore, bufferSize int, sync bool) Moni
 		store:    store,
 		handlers: handlers,
 		eventCh:  make(chan ConfigEvent, bufferSize),
-		sync:     sync,
+		sync:     syncMode,
+		lock:     &sync.RWMutex{},
 	}
 }
 
 func (m *configStoreMonitor) ScheduleProcessEvent(configEvent ConfigEvent) {
+	m.lock.Lock()
+	if m.closed.Load() {
+		m.lock.Unlock()
+		return
+	}
+
 	if m.sync {
+		m.lock.Unlock()
 		m.processConfigEvent(configEvent)
 	} else {
 		m.eventCh <- configEvent
+		m.lock.Unlock()
 	}
 }
 
+func (m *configStoreMonitor) waitAndCleanup(stop <-chan struct{}) {
+	<-stop
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.closed.Store(true)
+	// indicating the monitor has exited.
+	close(m.eventCh)
+}
+
 func (m *configStoreMonitor) Run(stop <-chan struct{}) {
 	if m.sync {
-		<-stop
+		m.waitAndCleanup(stop)
```

Comment:
This is too complex:

The target is to ensure all events processed before exit.

So for sync mod:  return immediately on stop


For async/buffer mode:  

close eventCh on stop in a go thread, and then

```
for ce := range m.eventCh {
		m.processConfigEvent(ce)
}
```

Reply:
Done, PTAL.